cranelift-codegen/src/verifier/mod.rs

```diff
@@ -468,6 +471,35 @@ impl<'a> Verifier<'a> {
         Ok(())
     }
 
+    /// Check that the given EBB can be encoded as a BB, by checking that only
+    /// branching instructions are ending the EBB.
+    fn encodable_as_bb(&self, ebb: Ebb, errors: &mut VerifierErrors) -> VerifierStepResult<()> {
+        // Skip this verification if the environment variable is not set.
+        if !self.verify_encodable_as_bb {
+            return Ok(());
+        };
+
+        let dfg = &self.func.dfg;
+        let inst_iter = self.func.layout.ebb_insts(ebb);
+        // Skip non-branching instructions.
+        let inst_iter = inst_iter.skip_while(|&inst| !dfg[inst].opcode().is_branch());
+        // Skip branching instructions.
+        let mut inst_iter = inst_iter.skip_while(|&inst| {
+            dfg[inst].opcode().is_branch() || dfg[inst].opcode().is_terminator()
```

Comment:
I think you should only accept at most one none-terminating branch and exactly one terminator -- that is pretty close to the final form we need.  The code above will accept multiple conditional exits.

In general even that test is too weak; we want to look for more specific patterns.  For example if the terminating instruction is BrTable you probably don't want to allow a conditional branch before it.  I actually suspect that we should only allow a conditional exit if the terminator is Fallthrough or Jump.  But it's possible that that is too hard to implement for the first iteration, so I won't insist on it.

Reply:
I changed the second `skip_while` by a manual iteration relying on the fact that an iterator is expected to return `None` when it finished and continues to return `None` on follow-up calls.